test/test_pebble.py

```diff
@@ -1076,6 +1161,423 @@ def test_get_services_names(self):
+            b"""
+--01234567890123456789012345678901
+Content-Disposition: form-data; name="files"; filename="/bad"
+
+bad path
+--01234567890123456789012345678901--
+""",
+        ))
+        with self.assertRaises(pebble.ProtocolError) as cm:
+            self.client.read_file('/etc/hosts', None)
+        self.assertEqual(str(cm.exception), "path not expected: /bad")
+
+        self.client.responses.append((
+            {'Content-Type': 'multipart/form-data; boundary=01234567890123456789012345678901'},
+            b"""
+--01234567890123456789012345678901
+Content-Disposition: form-data; name="files"; filename="/etc/hosts"
+
+bad path
+--01234567890123456789012345678901--
+""",
+        ))
+        with self.assertRaises(pebble.ProtocolError) as cm:
+            self.client.read_file('/etc/hosts', io.BytesIO())
+        self.assertEqual(str(cm.exception), 'no "response" field in multipart body')
+
+    def test_write_file_simple(self):
+        self.client.responses.append((
+            {'Content-Type': 'application/json'},
+            b"""
+{
+    "result": [
+        {"path": "/foo/bar"}
+    ],
+    "status": "OK",
+    "status-code": 200,
+    "type": "sync"
+}
+""",
+        ))
+
+        source = io.BytesIO(b'content')
+        self.client.write_file('/foo/bar', source)
+
+        self.assertEqual(len(self.client.requests), 1)
+        request = self.client.requests[0]
+        self.assertEqual(request[:3], ('POST', '/v1/files', None))
+
+        headers, body = request[3:]
+        content_type = headers['Content-Type']
+        dest = io.BytesIO()
+        req = self._parse_write_multipart(content_type, body, {'/foo/bar': dest})
+        self.assertEqual(dest.getvalue(), b'content')
+        self.assertEqual(req, {
+            'action': 'write',
+            'files': [{'path': '/foo/bar'}],
+        })
+
+    def test_write_file_options(self):
+        self.client.responses.append((
+            {'Content-Type': 'application/json'},
+            b"""
+{
+    "result": [
+        {"path": "/foo/bar"}
+    ],
+    "status": "OK",
+    "status-code": 200,
+    "type": "sync"
+}
+""",
+        ))
+
+        source = io.BytesIO(b'content')
+        self.client.write_file('/foo/bar', source, make_dirs=True, permissions=0o600,
+                               user='bob', group='staff')
+
+        self.assertEqual(len(self.client.requests), 1)
+        request = self.client.requests[0]
+        self.assertEqual(request[:3], ('POST', '/v1/files', None))
+
+        headers, body = request[3:]
+        content_type = headers['Content-Type']
+        dest = io.BytesIO()
+        req = self._parse_write_multipart(content_type, body, {'/foo/bar': dest})
+        self.assertEqual(dest.getvalue(), b'content')
+        self.assertEqual(req, {
+            'action': 'write',
+            'files': [{
+                'path': '/foo/bar',
+                'make-dirs': True,
+                'permissions': '600',
+                'user': 'bob',
```

Comment:
We don't seem to have a test of user-id/group-id testing along with the names. (I do think names are far more important, but if we are going to support uid/gid we should test it)

Reply:
Added a test of both kinds - thanks.